### DISCUSSION TITLE
RLS select is filtering rows by the column contact_deleted... so I can't so updates to that column?

### DISCUSSION THREAD

**Discussion by malalecherocks** [supabase/supabase#7908] [General]
Imagine we have a table called **contact** that has a column **contact_deleted** and in RLS SELECT we establish the rule that only allows contact_deleted = false

this way when our users do select it will only return rows of the contacts that haven't been deleted.

My problem is that when I try to do an UPDATE to establish the column contact_deleted to true, it gives me an error, it doesn't allow me to update it because as far as I read updates depends on selects.

I came to the conclusion that what I was trying to do is not possible, is that correct? Or maybe I was doing something wrong?

**Comment by GaryAustin1** [comment]
What are your update RLS policies?  You would need using policy (on the current row) to not check for contact_deleted.  The with check policy could be the same (it defaults to the same if not specified) as you are modifying the contact_deleted value.

If it is the select policy, then setting "returning minimal" would not do the select:
https://supabase.com/docs/reference/javascript/update#parameters

**Reply by GaryAustin1** [reply]
Did you try setting the returning minimal option in your update call?

**Reply by malalecherocks** [reply]
I'm about to put the kid to bed hopefully in 10 minutes I can try, I will let you know...

**Reply by malalecherocks** [reply]
I did:

```
const { data, error } = await supabase
				.from('team')
				.update({ deleted: true }, { returning: 'minimal' })
				.eq('id', $teamSelected.id);
```

I confirm my USING expression in SELECT is:
((is_member_of(uid(), id) OR (user_id = uid())) AND (deleted IS FALSE))

My UPDATE is the same in USING and in WITH CHECK:
true

And the error I receive is:
message: 'new row violates row-level security policy for table "team"'

If I remove from the policy SELECT the part of AND (deleted IS FALSE)
it works but it returns in every select all the teams, I mean for now what I do is a select filtering by that column, but it would be awesome to be able to only return the select of active teams...

**Reply by GaryAustin1** [reply]
I'm puzzled.  Update should not depend on select policy, especially with returning:'minimal'...

**Reply by GaryAustin1** [reply]
I'm a bit confused though, you talk about table contact, but your update is to 'team'?

**Reply by malalecherocks** [reply]
Yeah, my bad, I put it as a more simple question for the audience using the table "contact", but the reality is that the table is called team, and the user can delete a team, and what I do is setting a flag to deleted in that table. I read somewhere that you can't do updates if there is not a rls in select, so I understood that updates depends on rls select too.

I guess I could try with a function or trying to set up another flag to "execute_deletion" and then from the server doing the rest, which is putting the flag of deleted to 1, it's a huge workaround...

**Reply by GaryAustin1** [reply]
Deleting this...  I may have had the wrong key on for my test.  I'll reply again after I check.

**Reply by GaryAustin1** [reply]
After testing, update does not work if select policy is not met, independent of update policy and returning flag.
Docs should call that out in update, like insert.  Sorry for the chase.

https://github.com/PostgREST/postgrest/discussions/1844  After reading this seems everyone is confused.

Edit:  And this appears to be the answer from the Postgres docs....

> When multiple policies of different command types apply to the same command (for example, SELECT and UPDATE policies applied to an UPDATE command), then the user must have both types of permissions (for example, permission to select rows from the relation as well as permission to update them). Thus the expressions for one type of policy are combined with the expressions for the other type of policy using the AND operator.

**Reply by malalecherocks** [reply]
It makes sense that if you are going to update and the update itself needs to use a select under the hood it makes sense that the user should have permission to do that select. But my example goes further...

the thing is that that select we are trying to do: "deleted is not false" is met by the user, but the problem is that seems it's entering in a conflict with the update when the column we are trying to update its being update with the value "true" 

so weird... thanks anyway for taking a look at it, if anyone knows more about this please through some light

**Reply by GaryAustin1** [reply]
FWIW, I agree with your thinking.

**Reply by GaryAustin1** [reply]
So using an rpc call to a security definer function (bypasses RLS) to toggle your flag does not seem like a huge amount of work, versus 2nd flags and such.

**Reply by malalecherocks** [reply]
But if that RPC function bypass RLS anyone could call it and start setting that flag for all the rows… I think the solution I’m thinking could be the best would be…

for that specific update I call my api on my server, I check the user auth getting the user using:

** const { user } = await supabase.auth.api.getUserByCookie(req)**

although I read is having lots of issues specially with sveltekit since January.

once I have the user I can now confirm his user_id is authenticated and now I can just simoly bypass the RLS using the supabase object with the secret key.

is this correct? That’s what I was going to do today

**Reply by GaryAustin1** [reply]
In the rpc function you check using auth.uid() or auth.role for security how you like.   Similar to RLS then do the insert. It is a very common thing to do.
